lib/saml/url_service.rb

```diff
@@ -1,74 +1,109 @@
-      build_sso_url(authn_context: authn_context, relay_state: relay_state)
+    def idme_loa3_url
+      link_authn_context = authn_context.present? ? "#{authn_context}_loa3" : LOA::MAPPING.invert[3]
+      build_sso_url(link_authn_context)
     end
 
-    def mfa_url(current_user, relay_state)
-      policy = current_user.authn_context
-      authn_context = policy.present? ? "#{policy}_multifactor" : 'multifactor'
-      build_sso_url(authn_context: authn_context, relay_state: relay_state)
+    def mfa_url
+      link_authn_context = authn_context.present? ? "#{authn_context}_multifactor" : 'multifactor'
+      build_sso_url(link_authn_context)
     end
 
+    # SIGN OFF URLS
     # This is the internal vets-api url that first gets invoked, it should redirect without authentication
     # when this url gets invoked, the session should be destroyed, before the callback returns
-    def logout_url(session, relay_state)
+    def logout_url
       token = Base64.urlsafe_encode64(session.token)
-      Rails.application.routes.url_helpers.logout_v0_sessions_url(
-        success_relay: relay_state.relay_enum, session: token
-      )
+      Rails.application.routes.url_helpers.logout_v0_sessions_url(session: token)
     end
 
-    # SLO URLS
-    def slo_url(session, relay_state)
-      build_slo_url(session, relay_state)
+    def slo_url
+      logout_request = OneLogin::RubySaml::Logoutrequest.new
+      # cache the request for session.token lookup when we receive the response
+      SingleLogoutRequest.create(uuid: logout_request.uuid, token: session.token)
+      Rails.logger.info "New SP SLO for userid '#{session.uuid}'"
+      logout_request.create(url_settings)
     end
 
     private
 
     # Builds the urls to trigger various SSO policies: mhv, dslogon, idme, mfa, or verify flows.
-    # nil authn_context and nil connect will always default to idme level 1
-    # authn_context is the policy, connect represents the ID.me specific flow.
-    def build_sso_url(authn_context: LOA::MAPPING.invert[1], session: nil, relay_state: nil)
-      url_settings = url_settings(authn_context: authn_context, name_identifier_value: session&.uuid)
+    # link_authn_context is the new proposed authn_context
+    def build_sso_url(link_authn_context)
+      new_url_settings = url_settings
+      new_url_settings.authn_context = link_authn_context
       saml_auth_request = OneLogin::RubySaml::Authrequest.new
-      saml_auth_request.create(url_settings, RelayState: relay_state.login_url)
+      saml_auth_request.create(new_url_settings)
     end
 
-    # Builds the url to trigger SLO, caching the request
-    def build_slo_url(session, relay_state)
-      logout_request = OneLogin::RubySaml::Logoutrequest.new
-      Rails.logger.info "New SP SLO for userid '#{session.uuid}'"
+    def current_host
+      uri = URI.parse(saml_settings.assertion_consumer_service_url)
+      uri.to_s.gsub(uri.path.to_s + uri.query.to_s, '')
```

Comment:
Should this be `uri.path.to_s + '?' + uri.query.to_s`?

Reply:
how about `URI::HTTP.new(uri.scheme, nil, uri.host, uri.port, nil, nil, nil, nil, nil).to_s`?  gsub is the road to regret.